openedx/core/apidocs.py

```diff
@@ -0,0 +1,161 @@
+"""
+Open API support.
+"""
+
+import textwrap
+
+from drf_yasg import openapi
+from drf_yasg.generators import OpenAPISchemaGenerator
+from drf_yasg.utils import swagger_auto_schema as drf_swagger_auto_schema
+from drf_yasg.views import get_schema_view
+from rest_framework import permissions
+
+# -- Code that will eventually be in another openapi-helpers repo -------------
+
+
+class ApiSchemaGenerator(OpenAPISchemaGenerator):
+    """A schema generator for /api/*
+
+    Only includes endpoints in the /api/* url tree, and sets the path prefix
+    appropriately.
+    """
+
+    def get_endpoints(self, request):
+        endpoints = super(ApiSchemaGenerator, self).get_endpoints(request)
+        subpoints = {p: v for p, v in endpoints.items() if p.startswith("/api/")}
+        return subpoints
+
+    def determine_path_prefix(self, paths):
+        return "/api/"
+
+
+def dedent(text):
+    """
+    Dedent multi-line text nicely.
+
+    An initial empty line is ignored so that triple-quoted strings don't need
+    to start with a backslash.
+    """
+    if "\n" in text:
+        first, rest = text.split("\n", 1)
+        if not first.strip():
+            # First line is blank, discard it.
+            text = rest
+    return textwrap.dedent(text)
+
+
+def schema(
+    manual_parameters=None,
+    operation_summary=None,
+    operation_description=None
+):
+    """
+    Decorator for documenting an API endpoint.
+
+    This is heavily inspired from the the `drf_yasg.utils.swagger_auto_schema`__
+    decorator, but callers do not need to know about this abstraction.
+    All description fields should be in Markdown.
+
+    __ https://drf-yasg.readthedocs.io/en/stable/drf_yasg.html#drf_yasg.utils.swagger_auto_schema
+    """
+    for param in manual_parameters or ():
+        param.description = dedent(param.description)
+
```

Comment:
If you wanted to retain the no-parenthases `@api.schema` style, I think you could write something like
```python
def schema(
    manual_parameters=None,
    operation_summary=None,
    operation_description=None,
    *args,
):
    if args:
        provided_kwargs = (manual_parameters, operation_summary, operation_description)
        if len(args) > 1 or (not callable(args[0])) or any(provided_kwargs):
             raise ValueError("@openapi.schema takes no positional arguments; use keyword arguments instead.")
        return schema()(args[0])
```
Of course, you'd have to keep `provided_kwargs` up to date, which could be verified with a simple unit test.

Up to you whether it's worth the trouble!

Reply:
Yes, this is a possibility, but I'm not a big fan of this pattern; I think it's too convoluted and it makes the code much harder to understand. I'll keep the syntax for now.